### PR TITLE
not check initContianer status when all container status is running

### LINKED ISSUE #77859: Init Container will be removed by kubeGC when we set --maximum-dead-containers-per-container=0
**What happened**:
1.set --maximum-dead-containers-per-container=0
2.  create a deploy with a init container
3. watch the pod status , you will see the change 
```
[root@hpa-vm:~]$ kubectl get pod -w
NAME          READY   STATUS    RESTARTS   AGE
nginx-f8jfv   1/1     Running   0          30s
NAME          READY   STATUS     RESTARTS   AGE
nginx-f8jfv   0/1     Init:0/1   0          95s
nginx-f8jfv   1/1     Running    0          100s
```

**What you expected to happen**:
init container wont't be removed. 
   _**we need separate the containerGC and initContainerGC**_

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by lojies**
**What type of PR is this?**
 /kind bug

**What this PR does / why we need it**:
initContainers will be removed when --maximum-dead-containers or --maximum-dead-containers-per-container is satisfied.
Then initContainer's status will be nil, and initContainers will be think as not done. 
computePodActions  return changes and initContainer be recreated.
This may cause init setting error,because the pod has been running for a period and setting has been changed.
on the other hand, when the initContainer exited, it will be gc again and then be created again.
This process will be done again and again.

**Which issue(s) this PR fixes**:
Fixes  #77859

**Special notes for your reviewer**:
how i found this:
create pod by yaml
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
  labels:
    app: myapp
spec:
  containers:
  - name: myapp-container
    image: busybox
    command: ['sh', '-c', 'echo The app is running! && sleep 3600']
  initContainers:
  - name: init-myservice
    image: busybox
    command: ['sh', '-c', 'until nslookup myservice; do echo waiting for myservice; sleep 2; done;']
  - name: init-mydb
    image: busybox
    command: ['sh', '-c', 'until nslookup mydb; do echo waiting for mydb; sleep 2; done;']
```
```yaml
kind: Service
apiVersion: v1
metadata:
  name: myservice
spec:
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
---
kind: Service
apiVersion: v1
metadata:
  name: mydb
spec:
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9377
```
when gc happened, if the last initContainer init-mydb is removed, init-mydb is recreated.
if the initContainer init-myservice, it is ok.
**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 1]
```diff
@@ -500,10 +500,16 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 		}
 		return changes
 	}
-
+    allContainerStatusRunning := true
```
Comment (mattjmcnaughton) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Nit: Mind fixing the indentation here?

**Review by mattjmcnaughton — COMMENTED** [review]
/ok-to-test

Thanks for this PR! This seems like it fixes (or at least relates to) the issue raised in https://github.com/kubernetes/kubernetes/issues/77859. Mind updating your PR to reference it?

Could I ask you to please provide a unit test testing this behavior? Thanks :)

This overall behavior seems sensible to me - I'm interested in the opinions of @Pingan2017 .

**Comment by Pingan2017** [comment]
LGTM
Thanks, @lojies  Plz Run ./hack/update-gofmt.sh to fix the gofmt.

**Comment by Pingan2017** [comment]
iirc, when the init containers removed, the status of pod (by kubectl get pod) will trun to Init . 

https://github.com/kubernetes/kubernetes/blob/da31c50da191ae1c055069b936d9e549741e3e8a/pkg/printers/internalversion/printers.go#L579-L606

we also need solve this problem, if accept this.
/sig cli

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 1 above)
Comment (lojies) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go]: I have fixed the indentation and added a unit test for this behavior.
Please help me to do a review,thanks.

**Comment by sjenning** [comment]
The assumption in this PR is that if all of a pod's main containers are running, the init containers must have successfully executed in the past, even though the init containers might not exist in the runtime any more.

I've been trying to think though the container and pod level restart scenarios and I think this is a valid assumption.

In fact, we might even be able to assume that if *any* of the main containers are running, the init containers ran successfully in the past.

**Comment by lojies** [comment]
@sjenning @dchen1107 @mattjmcnaughton

**Pull request closed without merging**